### PR TITLE
835 callback functionality

### PR DESCRIPTION
**Pull request by melissapalmer**
**This PR introduces a Scheduler Task to Tilkynna which will, make a call to the given callback_url:** 

This task uses an internal DB queue (`_reports.callback_queue`), as its base for jobs. 

- When a report is requested the `_reports.generated_report` and `_reports.callback` are populated. 
- On report generation completion (whether SUCCESSFUL or FAILED), a row is saved to `_reports.callback_queue`
- the scheduled task will monitor this queue using postgres's `FOR UPDATE SKIP LOCKED` feature. 
- once the callback_url as been called or all retries events exhausted, `_reports.callback`  updated with final status and row is removed from `_reports.callback_queue` this helps to keep the queue table small 

image

### CODE REVIEW

File: doc/db/tilkynna.sql [hunk 1]
```diff
@@ -363,6 +364,43 @@ COMMENT ON COLUMN _reports.destination_parameter.validation IS 'RegEx used to va
...
+);
+-- ddl-end --
+COMMENT ON COLUMN _reports.callback_queue.retries_remaining IS 'Number of times left to rety upon failure ';
+-- ddl-end --
+ALTER TABLE _reports.callback_queue OWNER TO postgres;
+-- ddl-end --
+
+-- object: _reports.callback | type: TABLE --
+-- DROP TABLE IF EXISTS _reports.callback CASCADE;
+CREATE TABLE _reports.callback (
```
Comment (VictorSzewczenko) [inline doc/db/tilkynna.sql]: What is the intention of this table. The above one models the queue, does this one store results from callback delivery attempts?

File: src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java [hunk 2]
```diff
@@ -32,20 +32,20 @@
...
-            "       ORDER BY CASE d.downloadable WHEN true THEN 1 END, retry_count DESC, requested_at ASC LIMIT :batchSize FOR UPDATE SKIP LOCKED " + //
-            "   ) " + //
-            "   UPDATE  _reports.generated_report r " + //
-            "   SET    report_status = 'STARTED' " + //
-            "   FROM   jobs " + //
-            "   WHERE  r.correlation_id = jobs.correlation_id " + //
-            "   RETURNING CAST(r.correlation_id AS VARCHAR) ", nativeQuery = true)
-    public List<String> getBatchOfPendingGenerateReportJobs(@Param("batchSize") int batchSize);
-
+    // @Query(value = " WITH jobs AS ( " + //
```
Comment (VictorSzewczenko) [inline src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java]: Not sure if this should remain commented out?

File: doc/tilkynna_api.yml [hunk 3]
```diff
@@ -470,16 +470,16 @@ paths:
               schema:
                 $ref: '#/components/schemas/ErrorResponse'
       callbacks:
-        notificationEvent:
+        reportGenerationEvent:
```
Comment (VictorSzewczenko) [inline doc/tilkynna_api.yml]: Im wondering if webhooks are documented with OAS3 specs?

File: src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java (hunk 2 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/model/db/GeneratedReportEntityRepository.java]: This sorted out in https://github.com/GrindrodBank/tilkynna/compare/613ff014a4af...e9a6189d9df0

File: doc/tilkynna_api.yml (hunk 3 above)
Comment (melissapalmer) [reply doc/tilkynna_api.yml]: Yes it's here: https://swagger.io/docs/specification/callbacks/ 
- this is what has been used in that spec.

File: doc/db/tilkynna.sql (hunk 1 above)
Comment (melissapalmer) [reply doc/db/tilkynna.sql]: Yes,  this one stores the final results. 
In future could be enhanced with send again concepts similar to what github has.

File: src/main/java/org/tilkynna/report/callback/GenerateReportCallbackHandler.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @Autowired
+    private GeneratedReportCallbackProperties generatedReportCallbackProperties;
+
+    @Async("generateReportCallbackThreadPool")
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void run() {
+        log.debug("worker thread started. current thread name {}", Thread.currentThread().getName());
+
+        CallbackQueueEntity callbackQueueItem = callbackQueueEntityRepository.findCallbackQueueEntity(generatedReportCallbackProperties.getBackOffPeriodInMilliseconds());
+
```
[the remark this answers is not in the document]
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/callback/GenerateReportCallbackHandler.java]: This is taken care of by using postgres's `FOR UPDATE SKIP LOCKED` feature. 
- Which is the query running under the findCallbackQueueEntity function.

**Pull request merged**